[영화 티켓 예매 3, 4단계] 페토 미션 제출합니다.

chanho0908 opened the pull request:
# 셀프 체크리스트

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
-

## 코드 리뷰 커뮤니케이션

안녕하세요 페로로 ! 

3,4단계도 잘 부탁드립니다. MVP와 Mockk Test가 처음인지라 미숙한 부분이 많을 것 같습니다 🥲

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

<img src="https://github.com/user-attachments/assets/ccced5ba-a60b-47b0-a734-689cf3252078" width="300">

chanho0908 committed:
rename : 영화 목록 패키지명 수정

chanho0908 committed:
refactor : 러닝타임의 "분" 포맷을 strings resource로 분리 및 적용

chanho0908 committed:
feat : 포스터 리소스 객체 분리

chanho0908 committed:
refactor : Movie 객체 패키징 수정

chanho0908 committed:
build : mockk gradle dependecy 추가

chanho0908 committed:
test : Movie Fixture 추가

chanho0908 committed:
refactor : viewHolder 데이터 바인딩 함수에 수신 객체 지정 람다 with 적용

chanho0908 committed:
feat : 영화 목록 화면 MVP Pattern 적용 및 mockk 테스트 추가

- 테스트용 MovieModel 추가

chanho0908 committed:
rename : 예약 정보 도메인 패키징 수정

chanho0908 committed:
rename : 영화 목록 Presenter 패키지 수정

chanho0908 committed:
feat : 예매 화면 기본 UI 로직 MVP 적용

- 영화 정보 가져오기
- 인원수 증감

chanho0908 committed:
feat : 예약 완료 후 예약 내역 화면 이동 로직 MVP 적용 완료

chanho0908 committed:
refactor : 영화 정보를 예매 화면에 전달할 때 domain Model를 사용하지 않도록 수정

chanho0908 committed:
feat : 이미지가 없을 경우 placeholder 이미지를 보여주도록 구현

chanho0908 committed:
feat : 영화 상열일 출력 로직 MVP 적용 완료

chanho0908 committed:
feat : 영화 상영 시간 출력 로직 MVP 적용

chanho0908 committed:
feat : 날짜 선택시 시간 초기화 MVP 적용

chanho0908 committed:
feat : 예매 인원수 복원 기능 MVP 적용

chanho0908 committed:
file : 영화 포스터, 광고 이미지 추가

chanho0908 committed:
chore : 영화 일급 컬랙션, 포스터 이미지 객체 제거

chanho0908 committed:
feat : 영화 정보를 가져오는 데이터 소스 추가

chanho0908 committed:
feat : 영화 객체 Id 값 추가

chanho0908 committed:
feat : 광고 객체 추가

chanho0908 committed:
build : recyclerview dependency 추가

chanho0908 committed:
feat : 파일 네임으로 Drawble Resource Id를 가져오는 확장함수 추가

chanho0908 committed:
feat : 광고 뷰홀더, 레이아웃 추가

chanho0908 committed:
feat : 리사이클러뷰 & (광고, 영화) 멀티 뷰홀더 적용 및 UI Model 추가

chanho0908 committed:
feat : 영화 목록 화면 Presenter 기능 추가

- Domain Model -> Ui Model 매핑 함수 추가
- 10000개의 영화 데이터를 가져오는 로직 추가

chanho0908 committed:
refactor : domain -> UiModel 매핑 확장함수 위치 수정

chanho0908 committed:
feat : 예약 화면 Movie UIModel 적용 완료

chanho0908 committed:
refactor : 영화 저장소 객체 게터 오퍼레이터 함수 추가, 사이즈 반환 함수 제거

chanho0908 committed:
feat : espresso recyclerview test dependency 추가

chanho0908 committed:
feat : 영화 리사이클러뷰 화면 테스트 추가

chanho0908 committed:
docs : 기능 요구사항 업데이트

chanho0908 committed:
feat : 영화 객체의 ID를 영화 리스트 생성시 Index 값으로 수정

chanho0908 committed:
test : 영화 저장소 객체 생성시 테스트 누락 수정

chanho0908 committed:
refactor : 예약 프로퍼티 중 티켓 타입 제거

chanho0908 committed:
feat : 좌석 선택 layout 추가 및 intent 전달 테스트 완료

chanho0908 committed:
feat : 좌석 화면 뒤로가기 기능 추가

chanho0908 committed:
feat : 예매 좌석 도메인 모델 추가

chanho0908 committed:
feat : 테이블 레이아웃의 테로 사용될 좌표 객체 추가

chanho0908 committed:
feat : 좌선 "선택" 로직 MVP 적용

chanho0908 committed:
feat : 좌석 등급별 가격 정책 클래스 추가

chanho0908 committed:
feat : 좌석 선택시 예매 가능한 인원수 초과시 토스트 출력 기능 추가

- 좌표, 좌석 픽스처 추가

chanho0908 committed:
feat : 예매한 좌석 가격 출력 기능 추가

chanho0908 committed:
feat : 좌석을 예약 인원만큼 선택하지 않았을 때 버튼 비활성화 기능 추가

chanho0908 committed:
feat : 예약 완료시 예약 정보 전달 및 예약 완료 화면에서 예약 정보 출력 기능 추가

chanho0908 committed:
refactor : Presenter, Contract 패키지 분리

chanho0908 committed:
test : 불필요한 테스트 제거, 영화 더미 데이터 수정으로 인한 테스트 수정

chanho0908 committed:
refactor : 영화 목록화면에서 예매 화면 전환 로직 수정

chanho0908 committed:
rename : 좌석 화면 테스트 패키지 수정

chanho0908 committed:
chore : 불필요한 테스트 제거

chanho0908 committed:
feat : 좌석 선택후 예매 완료 버튼 클릭시 다이얼로그 출력 추가

chanho0908 committed:
refactor : 예매 화면에서 좌석 선택 화면 이동 전환 로직 수정

chanho0908 committed:
refactor : 좌석 선택 화면에서 예약 완료 화면으로 이동 로직 수정

chanho0908 committed:
refactor : 광고 출력 후 리사이클러뷰의 인덱스가 한개씩 밀려서 다음 영화가 출력되는 현상 수정

chanho0908 committed:
refactor : 영화 전체 리스트 반환시 새로운 컬렉션을 깊은 복사해 반환하도록 수정

chanho0908 committed:
feat : 영화관 좌석수 만큼 예매 인원 제한 기능 추가

chanho0908 committed:
refactor : 좌석 선택 화면 레이아웃 수정

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/seat/SeatContract.kt:
```diff
@@ -0,0 +1,42 @@
+package woowacourse.movie.presenter.seat
```
view, presenter라는 패키지가 왜 분리가 된건가요?
SeatXXX를 한곳에서 모아서 관리하는 것 대비 어떤 이점이 있는지 궁금합니다.

chanho0908 replied:
`Presenter`는 `Model`과 `View`의 중간 다리 역할을 하게되는데
이 때문에 View에 Presenter가 위치하는 것이 어색 하다고 느껴 분리하였습니다 !

다만 리팩토링 과정에서 패키지가 분리되어 작업을 하는데 있어 같은 패키지에 위치했을 때 보다 불편함이 다소 발생한점,

`View`는 `Presenter`에 대한 강한 의존성을 가지기 때문에 `Presenter`를 별도 패키지로 분리하기보다

`View`와 `Presenter`를 같은 패키지에 묶어 `View` 중심으로 관리하는 편이 구조적으로 자연스럽다고 생각해 같은 패키지로 수정하겠습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/seat/SeatContract.kt:
```diff
@@ -0,0 +1,42 @@
+
+interface SeatContract {
+    interface View {
+        fun onClickSeat(
+            position: Coordination,
+            peopleCount: Int,
+        )
```
Presenter가 View에게 클릭을 하라는 요청을 할 경우가 있나요?

chanho0908 replied:
이벤트를 트리거시키고 사용자와 상호작용하는 것은 View의 책임입니다.

아직 MVP에 대한 지식이 부족해 발생한 문제인 것 같습니다 😅

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/seat/SeatContract.kt:
```diff
@@ -0,0 +1,42 @@
+            limit: Int,
+        )
+
+        fun onConfirmClicked(limit: Int)
```
click이라는 구체적인 동작은 View의 관심사입니다.
가령 더블클릭이나 롱클릭 드래그로 이벤트가 시작된다면 인터페이스 이름도 변경됩니다.

View와 Presenter의 인터페이스는 기능 관점에서 작명 해주시는 것이 좋습니다.

chanho0908 replied:
말씀해주신 리뷰에 대한 고민을 피드백 문서 6-3번 항목에 작성하였습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/seat/SeatContract.kt:
```diff
@@ -0,0 +1,42 @@
+        fun onConfirmClicked(limit: Int)
+    }
+
+    companion object PresenterFactory {
+        fun providePresenter(view: View): Presenter {
+            val seats = Seats()
+            return SeatPresenter(view, seats)
+        }
+    }
```
Factory를 왜 인터페이스가 구현하는지 모르겠어요.
기껏 인터페이스로 만들어서 구현체 의존성 모르도록 분리했는데 다시 결합시키면 무슨 소용인가요?

chanho0908 replied:
앗 .. 자꾸만 인터페이스에서 구현체를 알게하는 실수를 하네요 

오목 미션에서도 비슷한 문제로 DIP를 위반하는 문제가 있었는데 😭 수정하도록 하겠습니다

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/seat/SeatPresenter.kt:
```diff
@@ -0,0 +1,77 @@
+        limit: Int,
+    ) {
+        val newSeat =
+            Seat(
+                x = position.x.value,
+                y = position.y.value,
+            )
```
Position으로 x, y를 Wrapping했는데 Seat는 다시 x, y를 넣어주는 이유가 있나요?

chanho0908 replied:
`coordination` 은 UI에서 사용될 좌표값이고 Seat는 도메인의 비즈니스 로직을 사용하기 위한 
모델이기 때문에 이 둘을 분리하기 위함입니다 !

laco-dev replied:
Coordination이나 Row, Column이나 모두 비슷한데 UI에서만 써야할 정도로 UI 의존성이 강한건가요?
UI는 도메인 모델을 쓰면 안되는건지?

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/seat/SeatPresenter.kt:
```diff
@@ -0,0 +1,77 @@
+                y = position.y.value,
+            )
+
+        when (seats.isSelected(newSeat)) {
+            true -> seats.removeSeat(newSeat)
+            false -> {
+                if (canSelect(limit).not()) return
+                seats.addSeat(newSeat)
+            }
+        }
```
이 로직은 모두 비즈니스 로직 같네요. Seats가 해야하지 않을까요?

```kotlin
seats.toggleSeat(coordinates)
````

chanho0908 replied:
동의합니다 ! 

선택된 좌석을 추가, 제거하는 과정에서 자신의 상태에 따라 분기하고, 
그 결과에 따른 책임을 수행하는 것은 Seats 스스로가 판단하고 수행해야할 책임인 것 같습니다 

또한 `isSelected ` 와 `canSelect `을 `private`으로 만드는 이점도 생길 것 같습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/seat/SeatPresenter.kt:
```diff
@@ -0,0 +1,77 @@
+        view.setConfirmButtonEnabled(isEnabled)
+    }
+
+    private fun canSelect(limit: Int): Boolean {
+        if (seats.canSelect(limit).not()) {
+            view.showToast(limit)
+            return false
+        }
+        return true
+    }
```
토스트까지 출력하면 2가지 일을 하고 있지 않나요?

chanho0908 replied:
맞습니다 🥲 선택 가능 여부와 토스트 출력 두 가지 책임을 가지고 있네요

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/seat/SeatPresenter.kt:
```diff
@@ -0,0 +1,77 @@
+            }
+    }
+
+    private fun seatToLabel(): String {
+        return seats
+            .item
+            .joinToString {
+                val rowLetter = ('A' + it.x - 1)
+                val columnNumber = it.y
+                "$rowLetter$columnNumber"
+            }
+    }
```
'A' 라고 출력하는건 프레젠터의 관심사인가요? UI의 관심사인가요?

chanho0908 replied:
제가 생각하기에 `Presenter`의 책임 중 하나가 `Domain`을 Ui에 보여줄 형태로 매핑하는 것이라고 생각 했습니다.

하지만 그 동안 1단계 미션을 해왔던 것을 되돌아봤을 때,
Ui에 보여줄 형태로 변형하는 것은 `View`의 역할인 것으로 이해해 왔기에 Presenter에 존재하는 매핑 로직을 제거 하겠습니다.

laco-dev replied:
매핑하는것도 맞는 말이지만 UI에서 사용할 모델로 바꾸는 것과 실제 UI로 바꾸는 것은 차이가 있습니다,

"a1'이라고 되어있는 "이미지"를 변환하는 것도 프레젠터가 해야하는건가요?
만약 디자이너가 A1이라고 출력되어있는거 'a1'으로 바꿔주세요. 하면 프레젠터 테스트를 다시 짤건가요..?

laco-dev commented on app/src/main/java/woowacourse/movie/view/seat/SeatActivity.kt:
```diff
@@ -0,0 +1,185 @@
+
+class SeatActivity : AppCompatActivity(), SeatContract.View {
+    private lateinit var presenter: SeatContract.Presenter
+    private lateinit var defaultBooking: Booking
```
defaultBooking을 액티비티가 가지고 있는게 맞나요? 프레젠터가 가지고 있는게 맞나요?

chanho0908 replied:
이 또한 하나의 Domain이기 때문에 프레젠터가 가지고 있는게 맞다고 생각합니다.

laco-dev commented on app/src/main/java/woowacourse/movie/view/seat/SeatActivity.kt:
```diff
@@ -0,0 +1,185 @@
+        presenter.changeSeat(position, peopleCount)
+    }
+
+    override fun showSeat(seat: List<Coordination>) {
+        this.seat.children
+            .filterIsInstance<TableRow>()
+            .forEach { row ->
+                row.children
+                    .filterIsInstance<TextView>()
+                    .forEach { textView ->
+                        val seatCoord = textView.tag as? Coordination
+                        if (seatCoord != null && seat.contains(seatCoord)) {
+                            textView.setBackgroundColor(Color.YELLOW)
+                        } else {
+                            textView.setBackgroundColor(0)
+                        }
+                    }
+            }
+    }
```
시트를 보여줄 때 마다 매번 View를 다시 찾으면 꽤나 번거로워 보이네요.
함수의 depth도 많이 깊어진 상태라 개선해볼 수 있다면 좋겠어요

laco-dev commented on app/src/main/java/woowacourse/movie/view/seat/SeatActivity.kt:
```diff
@@ -0,0 +1,185 @@
+        seats: String,
+        price: Int,
+    ) {
+        val intent =
+            BookingCompleteActivity.newIntent(
+                this,
+                defaultBooking.title,
+                defaultBooking.bookingDate,
+                defaultBooking.bookingTime,
+                defaultBooking.count.value,
+                price,
+                seats,
+            )
```
newIntent의 파라미터로 Booking을 넘겨주면 되지 않을까요?

chanho0908 replied:
예약 완료 화면에선 전달받은 데이터를 화면에 보여주는 일만 하기 때문에
원시값을 전달해주면 Presenter를 사용할 필요가 없어질 것이라고 생각했는데,

오히려 전달하는 파리미터가 늘어나면서 코드가 지저분해진 것 같습니다

크루들한태 보여주니 다들 기겁을 하더군요 😣

최종 예약 결과를 보여기 위한 `Ticket`이란 객체를 만들어 깔끔하게 정리했습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/booking/BookingContract.kt:
```diff
@@ -0,0 +1,71 @@
+import woowacourse.movie.view.movies.model.UiModel
+import java.time.LocalDateTime
+
+interface BookingContract {
```
왜 Booking과 BookingComplete가 Contract를 재사용하네요. 😅 
MVP 패턴에서는 코드가 조금이라도 겹치면 Contract를 재사용하도록 되어 있나요?

chanho0908 replied:
MVP의 단점 중 하나가 Presnter를 재사용 할 수 없다는 점이라고 알고있는데,

이 말은 Contract 또한 특정 View의 기능과 역할에 의존한다고 생각하여

BookingComplete 화면에서도 별도의 Contract를 사용하도록 수정했습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/booking/BookingContract.kt:
```diff
@@ -0,0 +1,71 @@
+
+        fun showToast()
+
+        fun onClickIncrease()
+
+        fun onClickDecrease()
+
+        fun onClickBooking()
+
+        fun onSelectDate(selectedDate: String)
```
마찬가지로 onXXX은 프레젠터가 View에게 호출하는 것이 아니므로 모두 제거해주세요

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/booking/BookingContract.kt:
```diff
@@ -0,0 +1,71 @@
+
+        fun moveToBookingComplete(booking: Booking)
+
+        fun restoreSavedState(savedCount: Int)
```
restoreSavedState동작도 액티비티의 문제이기 때문에 View의 인터페이스로 보이지 않습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/booking/BookingPresenter.kt:
```diff
@@ -0,0 +1,90 @@
+        view.showPeopleCount(count.value)
+    }
+
+    override fun loadScreeningDate(
+        startDate: String,
+        endDate: String,
+        now: LocalDateTime,
+    ) {
```
불필요한 포매팅을 프레젠터가 알아야 합니다.
View로부터 StartDate, EndDate를 LocalDate 타입으로 받으면 되는것 아닌가요?

laco-dev commented on app/src/main/java/woowacourse/movie/view/movies/model/UiModel.kt:
```diff
@@ -0,0 +1,38 @@
+import woowacourse.movie.view.StringFormatter
+import woowacourse.movie.view.movies.model.UiModel.MovieUiModel
+
+sealed interface UiModel {
+    data class MovieUiModel(
+        val id: Int,
+        val title: String,
+        val imgName: String,
+        val releaseStartDate: String,
+        val releaseEndDate: String,
+        val runningTime: Int,
+    ) : UiModel
+
+    data class AdvertiseUiModel(
+        val imgResource: String,
+    ) : UiModel
+}
```
지금은 단순히 모든 정보가 똑같을거라고 가정하고 UiModel을 여러곳에서 쓰고있어요.
하지만 지금 구조는 영화, 광고를 분리해서 보여주기 위한 아이템 전략이지 여러 화면에서 재사용 하기 위한 구조로 보기는 어렵습니다.

Booking 화면에서는 다르게 출력해야 한다거나 불필요한 광고 관련 정보도 알고 있는 등 의존성이 만들어집니다.
저라면 차라리 Movie 모델을 그대로 전달하고 재사용 되는 UI 로직을 어떻게 처리할지 고민하는 방법이 나아 보입니다.

chanho0908 replied:
말씀하신 내용에 적극 공감합니다.

스스로도 리사이클러뷰에 사용된 모델을 재활용 하는 것에 대해서 의문은 들었지만 그 의문에 대한 근거를 찾지 못했는데
페로로께서 말씀해주신 것을 고민해보니 이 의문에 대한 근거를 찾은 것 같습니다 😊

laco-dev commented on app/src/main/java/woowacourse/movie/view/movies/viewholder/MovieViewHolder.kt:
```diff
@@ -0,0 +1,47 @@
+    private val movieRunningTime = itemView.findViewById<TextView>(R.id.tv_running_time)
+    private val bookingBtn = itemView.findViewById<Button>(R.id.btn_booking)
+
+    init {
+        bookingBtn.setOnClickListener { onClickBooking(bindingAdapterPosition) }
+    }
```
클릭 리스너로 Movie를 그대로 돌려주면 되는데 굳이 Int로 넘겨주어서 사용하는 곳을 불편하게 만들 필요가 있나요?
그거 이상으로 init에서 리스너를 사용하는게 훨씬 유용하다고 보시나요?

chanho0908 replied:
이 부분은 수정하면서 [문서](https://github.com/woowacourse/android-movie-ticket/pull/127/files#diff-f5630abb6b1d8c90ac6998aeb256e9d012aede469eac1f482738913cf0903020)를 작성했는데, 1번 항목을 참고 부탁드립니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/view/movies/adapter/MovieAdapter.kt:
```diff
@@ -0,0 +1,53 @@
+) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
+    override fun getItemCount(): Int = itemsList.size
+
+    override fun getItemViewType(position: Int): Int =
+        when (itemsList[position]) {
+            is UiModel.MovieUiModel -> VIEW_TYPE_MOVIE
+            is UiModel.AdvertiseUiModel -> VIEW_TYPE_ADVERTISEMENT
+        }
```
sealed class를 사용하셨다면 ListAdapter에 대해 들어보셨을거라고 생각합니다.
기회가 된다면 ListAdapter에 대해 학습 해보시면 됩니다.

chanho0908 replied:
제가 생각하기에 리사이클러뷰 어댑터가 아닌 리스트뷰 어댑터를 사용하는 상황은
리사이클러뷰 아이템을 수정해야하는 일이 발생해 `notifiyDataSetChange` 같은 메서드를 사용할 때 라고 생각합니다.

아직은 리사이클러뷰의 전달된 데이터가 변경될 일이 없어 리사이클러뷰를 사용하였습니다 !

추후 리사이클러뷰의 아이템을 추가/삭제/수정 등의 요구 사항이 생긴다면 학습해보겠습니다 ☺️

laco-dev commented on app/src/main/java/woowacourse/movie/view/movies/MovieListActivity.kt:
```diff
@@ -0,0 +1,52 @@
+        presenter.loadUiData()
+    }
+
+    override fun showMovieList(movieList: List<UiModel>) {
+        val rv = findViewById<RecyclerView>(R.id.rv)
+        val adapter =
+            MovieAdapter(
+                itemsList = movieList,
+                onClickBooking = { presenter.onSelectMovie(it) },
+            )
+        rv.layoutManager = LinearLayoutManager(this)
+        rv.adapter = adapter
```
MovieList를 보여줄 때 마다 어댑터를 새로 만들면 어떤 문제가 있나요?
가령 영화 목록이 페이지로 되어 있어 아이템을 뒤에 계속해서 추가하는 구조라고 생각해보세요

chanho0908 replied:
지금은 최초에 영화 리스트를 한 번만 보여주기 때문에 문제가 없지만,
말씀하신 것 처럼 만약 영화 목록이 페이징  구조라면,
showMovieList()가 호출될 때마다 어댑터를 새로 생성하게 됩니다.

이 경우 발생할 수 있는 문제는 다음과 같습니다:

1. RecyclerView의 스크롤 성능 저하

기존 어댑터와 연결된 ViewHolder들을 재사용할 수 없어지기 때문에,
매번 새로운 어댑터가 생성되고, RecyclerView가 전체 아이템을 다시 그리게 됩니다.

이 과정에서 정크 라고 하는 버벅임 현상이 일어납니다.

2. 불필요한 리소스 낭비

Adapter 객체를 매번 새로 생성하면 메모리 사용량이 증가하고, 불필요한 Garbage Collection이 유발될 수 있습니다.

따라서 일반적으로는 Adapter는 최초에 한 번만 생성하고,
데이터가 변경될 때는 ListAdapter를 사용하는 방식으로 갱신하는 것이 올바른 패턴이라고 생각합니다.

laco-dev commented on app/src/main/java/woowacourse/movie/view/seat/SeatActivity.kt:
```diff
@@ -0,0 +1,185 @@
+package woowacourse.movie.view.seat
```
여기는 스크린샷이 없네요 🙀

chanho0908 replied:
앗... 스크린삿이 짤려있네요 잘 확인했어야하는데... 다시 올리겠습니다 !

laco-dev reviewed (changes requested):
3, 4단계 고생 많으셨습니다.

MVP에서 각 컴포넌트들이 하는 역할과 관심사에 대해 개념을 잘 잡아두셔야 합니다.
이 내용 위주로 의견을 남겨드렸으니 하나만 보고 하나만 고치지 말고 비슷한 다른 문제가 있는지 다시 전반적으로 돌아보시면 됩니다.

chanho0908 committed:
feat : 문서용 디렉토리 추가

chanho0908 committed:
refactor : 리사이클러뷰의 리스너 인덱스 반환으로 인한 문제 해결 및 영화 데이터 접근 최적화

chanho0908 committed:
chore : 기타 코드 개선

- 예약 완료 화면 패키지 분리
- 좌석 테이블 레이아웃의 셀 높이가 0으로 되어 좌석이 보이지 않던 문제 수정
- 사용하지 않는 AbstractListViewAdapter class 제거

chanho0908 committed:
refactor : Presenter의 DIP 위반 수정

chanho0908 committed:
refactor : Presenter의 비즈니스 로직 제거

- view가 초기화 됐을 때 0원을 보여주기 위한 호출 추가

chanho0908 committed:
refactor : Presenter의 UI 로직 제거

chanho0908 committed:
refactor : 테이블 레이아웃 코드 개선

chanho0908 committed:
refactor : Prestener에서 비즈니스 로직을 요청 하지 않는 메서드 제거

chanho0908 committed:
refactor :  Presenter 함수 네이밍 수정

chanho0908 committed:
refactor : 좌석 화면에서 intent로 전달 받은 기본 예약 정보를 Presenter가 관리하도록 수정

chanho0908 committed:
feat : 예약 완료 화면 MVP 적용

chanho0908 committed:
refactor : 좌석 예매 -> 예매 완료 화면에 티켓 전달시 좌석 리스트를 전달하도록 수정

chanho0908 committed:
rename : presenter, contract 패키지 수정

chanho0908 committed:
refactor : 광고 레이아웃 수정

chanho0908 commented on app/src/main/java/woowacourse/movie/domain/model/seat/Seats.kt:
```diff
@@ -0,0 +1,35 @@
+                true
+            }
+        }
+    }
```
https://github.com/woowacourse/android-movie-ticket/pull/127#discussion_r2062001014

해당 코멘트 적용했습니다.

chanho0908 commented on app/src/main/java/woowacourse/movie/view/seat/SeatRow.kt:
```diff
@@ -0,0 +1,36 @@
+    private val row: TableRow,
+    private val rowIndex: Int,
+    private val onSeatClick: (Coordination) -> Unit,
+) {
```
https://github.com/woowacourse/android-movie-ticket/pull/127#discussion_r2062029818

해당 코멘트를 적용하고자 테이블을 각각의 View와 Row로 나누었습니다.

chanho0908 reviewed (commented):
안녕하세요 페로로 !

MVP에 대한 지식이 많이 부족했지만 페로로의 리뷰 덕에 감을 잡을 수 있었던 것 같습니다

1,2 단계 미션에서 제 생각을 페로로께 전달하는 과정에서 제 설명이 다소 부족했다고 느껴

어떻게 해야 제 생각을 전달할 수 있을까 고민하다 리팩터링 과정에서 고민한 문제들을
별도의 문서로 정리해 보았습니다 !

이걸로 제 생각이 잘 전달 되었으면 좋겠네요 ☺️

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/booking/BookingContract.kt:
```diff
@@ -24,7 +26,7 @@ interface BookingContract {
         fun onClickBooking()
 
-        fun onSelectDate(selectedDate: String)
+        fun onSelectDate(selectedDate: LocalDate)
```
반영이 안된 것 같아요.
View.onClickBooking이라고 프레젠터가 호출할 일이 없는데 왜 인터페이스로 있어야 하나요?

chanho0908 replied:
이 부분도 수정되었습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/booking/BookingPresenter.kt:
```diff
@@ -5,49 +5,44 @@ import woowacourse.movie.domain.model.booking.Booking
-    override fun loadMovieDetail(index: Int) {
-        view.showMovieDetail(movies[index].toUiModel())
+    override fun loadMovieDetail(id: Int) {
+        view.showMovieDetail(movies[id])
```
참고)
프레젠터를 lazy하게 초기화 한다면 생성자로부터 movieId를 전달받는 전략도 사용할 수 있습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/view/booking/BookingActivity.kt:
```diff
@@ -45,37 +48,37 @@ class BookingActivity : AppCompatActivity(), BookingContract.View {
-            presenter.loadScreeningDate(releaseStartDate, releaseEndDate, LocalDateTime.now())
-            presenter.loadScreeningTime(dateSpinner.selectedItem.toString(), LocalDateTime.now())
+            presenter.loadScreeningDate(startDate, endDate, LocalDateTime.now())
+            presenter.loadScreeningTime(dateSpinner.selectedItem as LocalDate, LocalDateTime.now())
```
showMovieDetail을 호출하는 시점에 ScreeningDate ScreeningTime을 보여주도록 같이 전달하기는 어려운가요?
Presenter -> View -> Presenter -> View로 움직이고 있어서요

```kotlin
// Presenter
view.showMovieDetail(...)
view.showScreeningDateTime(...)
```

chanho0908 replied:
https://github.com/woowacourse/android-movie-ticket/pull/127/commits/d7b66adf9c46f76574b7b8cd478d8f9aadedb09d

해당 코멘트 적용했습니다 !

시간 스피너는 날짜 스피너에 의존한다고만 생각에 갇혀 말씀하신 방법을 생각해내지 못했네요 🥲

laco-dev commented on app/src/main/java/woowacourse/movie/view/movies/viewholder/MovieViewHolder.kt:
```diff
@@ -13,6 +13,7 @@ class MovieViewHolder(
     itemView: View,
     onClickBooking: (Int) -> Unit,
 ) : RecyclerView.ViewHolder(itemView) {
+    private var mId: Int = NO_ID
```
`mId` 😠

chanho0908 replied:
하하핫 ... 😅

laco-dev commented on app/src/main/java/woowacourse/movie/view/movies/viewholder/MovieViewHolder.kt:
```diff
@@ -22,11 +23,12 @@ class MovieViewHolder(
 
     init {
-        bookingBtn.setOnClickListener { onClickBooking(bindingAdapterPosition) }
+        bookingBtn.setOnClickListener { onClickBooking(mId) }
```
> 1. 리사이클러뷰의 리스너 인덱스 반환으로 인한 문제 해결 및 영화 데이터 접근 최적화

저는 그럼에도 아직 아쉽다고 생각하는데요.
명확하게 어떤 Movie를 클릭했는지 알 수 있음에도 불구하고 id를 전달하는게 유리하다고 생각하시나요?
만약 아이디가 없는 아이템을 써야 하면 어떻게 할지 궁금합니다.

chanho0908 replied:
이 문제에 대해서 크루들과 토론을 했습니다.

그 결과 현재 MovieUiModel의 정보들로 Movie를 만들어서 반환하자는 의견을 들었고 
저는 페로로께서 말씀하신 방법이 이 방법이 아닐까 싶어 적용해봤습니다 !

그리고 피드백 수업에서도 `init`에서 리스너를 초기화 하는 방법을 배웠습니다.

저 또한 여러 아티클을 보고 공부하면서 init 에서 리스너를 바인딩하는 것을 정답처럼 여겨왔는데요,
이 과정에서 느낀점은 사용하는 곳이 어려워진다 였습니다.

예컨데, 지금처럼 멀티 뷰홀더를 사용할 경우 포지션에 해당하는 List의 원소를 찾는 계산 코드가 추가되는 문제도 발생합니다.
늘 그렇듯, 모든 것은 트레이드 오프 관계라고 생각해요!

그럼 그 안에서 스스로 중요하게 여기는 요소들과 이유들을 찾는것이 정답일 것 같습니다

chanho0908 replied:
https://github.com/woowacourse/android-movie-ticket/pull/127/commits/fabb305e0bf555ea1f3d9121d0e029003006b2c5

해당 코멘트 적용했습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/presenter/seat/SeatPresenter.kt:
```diff
@@ -10,30 +10,20 @@ class SeatPresenter(
+            view.showPrice(seats.bookingPrice())
+            updateConfirmButtonState(limit)
+        } else {
+            view.showToast(limit)
```
조금 더 간결하게 작성한다면 이렇게도 할 수 있겠네요.
```kotlin

if (!seats.canSelect(newSeat, limit) {
  return view.showToast(limit)
}

seats.toggleSeat(...)
view.xxx
```

chanho0908 replied:
https://github.com/woowacourse/android-movie-ticket/pull/127/commits/e4cc385c84cbb837890e99143a1435762d74e9ac

해당 코멘트 적용했습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/view/seat/SeatView.kt:
```diff
@@ -0,0 +1,32 @@
+) {
+    private val rows = mutableListOf<SeatRow>()
+
+    fun initSeats() {
+        seatTable.children
+            .filterIsInstance<TableRow>()
+            .forEachIndexed { rowIndex, row ->
+                val rowManager = SeatRow(row, rowIndex, onSeatClick)
+                rowManager.initSeats()
+                rows.add(rowManager)
+            }
+    }
```
좋습니다 👍🏻 

더 고도화를 한다면 TabLayout 하나만 만들고 TableRow와 TextView는 코드로 생성하여 완전히 동적으로 테이블을 편성하는 방법이 있겠네요.

동적으로 뷰를 생성하는건 학습 범위와 넘어가는 주제라 참고만 해주세요

chanho0908 replied:
이 부분도 고려를 했었는데 시간상의 이유로 적용을 못했습니다 감사합니다 페로로 ☺️

laco-dev commented on app/src/main/java/woowacourse/movie/view/booking/BookingActivity.kt:
```diff
@@ -45,7 +45,9 @@ class BookingActivity : AppCompatActivity(), BookingContract.View {
         initView(movieIdx)
         savedInstanceState?.let {
-            restoreSavedState(it)
+            presenter.restorePeopleCount(it.getInt(KEY_PEOPLE_COUNT))
+            val savedTimePosition = it.getInt(KEY_SELECTED_TIME_POSITION)
+            timeSpinner.setSelection(savedTimePosition)
```
이것도 비슷하게 presenter 초기화 하는 시점에 넣어줄 수 있다면 더 쉽게 처리할 수 있겟네요.

laco-dev commented on app/src/main/java/woowacourse/movie/view/complete/BookingCompleteActivity.kt:
```diff
@@ -11,30 +11,38 @@ import androidx.core.view.ViewCompat
             insets
         }
         supportActionBar?.setDisplayHomeAsUpEnabled(true)
+        presenter.loadTicket()
```
이거도 init 에서 같이 호출해주면 될 것 같네요

laco-dev commented on app/src/test/java/woowacourse/movie/presenter/SeatPresenterTest.kt:
```diff
@@ -0,0 +1,78 @@
+            count = PeopleCount(2),
+        )
+
+    @BeforeEach
+    fun setUp() {
+        mockView = mockk(relaxed = true)
+        mockSeats = mockk(relaxed = true)
+        presenter = SeatPresenter(mockView, mockSeats, booking)
+    }
```
Mockk을 잘 활용했네요. 👍🏻 

프레젠터 입장에서는 구체적인 비즈니스 로직이 어떻게 나올지 관심 없으니 적절한 given을 준비하는 걸로 충분합니다.
마이너한 내용이지만 mockk은 생성 시점에 한번만 해주셔도 됩니다.
```kotlin
private val mockSeats: Seats = mockk(...)
```

chanho0908 replied:
이런 꿀팁이 ! 😱 감사합니다 페로로 !

laco-dev commented on app/src/test/java/woowacourse/movie/presenter/MovieListPresenterTest.kt:
```diff
@@ -0,0 +1,43 @@
+
+        presenter.loadUiData()
+
+        verify {
+            view.showMovieList(
+                match { uiModels ->
+                    uiModels[0] is UiModel.MovieUiModel &&
+                        uiModels[1] is UiModel.MovieUiModel &&
+                        uiModels[2] is UiModel.MovieUiModel &&
+                        uiModels[3] is UiModel.AdvertiseUiModel
+                },
+            )
+        }
```
각각의 uiModels를 검사하는 것도 좋지만 기댓값을 미리 만들어두고 비교하는 것도 좋을 것 같습니다.

```kotlin
val expected = UiModel(...)
verify { view.showMovieList(expected) }
```

chanho0908 replied:
https://github.com/woowacourse/android-movie-ticket/pull/127/commits/f371ed6503b3bec855b0a0729666ff5e82c34fd1

해당 코멘트 적용했습니다 !

laco-dev commented on app/src/test/java/woowacourse/movie/presenter/BookingPresenterTest.kt:
```diff
@@ -0,0 +1,177 @@
+
+        // then
+        verify {
+            view.showMovieDetail(
+                match {
+                    it.id == movie.id &&
+                        it.title == movie.title &&
+                        it.posterResource == movie.posterResource &&
+                        it.releaseDate == movie.releaseDate &&
+                        it.runningTime == movie.runningTime
+                },
+            )
```
여기도 Movie를 미리 정의해두고 그 값을 재사용하면 될 것 같아요

chanho0908 replied:
https://github.com/woowacourse/android-movie-ticket/pull/127/commits/f371ed6503b3bec855b0a0729666ff5e82c34fd1

해당 코멘트 적용했습니다.

laco-dev commented on app/src/test/java/woowacourse/movie/presenter/BookingPresenterTest.kt:
```diff
@@ -0,0 +1,177 @@
+    fun `인원이 1명 증가한다`() {
+        // given
+        var count = 1
+        every { view.showPeopleCount(count) } just Runs
```
relaxed로 처리하면 생략해도 될까요?

chanho0908 replied:
앗 이미 `relaxed` 로 기본 동작을 하게끔 구현했는데 불필요한 구문이 또 들어갔군요 !

laco-dev reviewed (changes requested):
피드백 반영 고생 많으셨습니다.

이모지로만 답변 남겨주시면 히스토리 파악이 어려워서 가급적이면 답변을 같이 남겨주시면 좋겠습니다.

추가로 고려해볼만한 내용과 테스트 관점에서 팁 몇가지 남겨두었어요.

chanho0908 committed:
refactor : Seat와 중복되는 Ui 용 좌표 Model 제거

chanho0908 committed:
refactor : 영화 목록 리사이클러뷰 선택시 Movie 객체를 반환하도록 수정

chanho0908 committed:
refactor : 좌석 선택/해제 판별 함수 수정

chanho0908 committed:
test : 테스트 코드 수정

- mockk 객체는 한번만 초기화하도록 수정
- expected 객체 분리

chanho0908 committed:
refactor : 영화 정보 초기화시 Presenter에서 시간 날짜 정보도 초기화하도록 수정

- 테스트용 날짜 Fixture 추가

chanho0908 committed:
refactor : 예매 완료 화면 Presenter 초기화시 예매 정보를 호출하도록 수정

- 예매 완료 화면 Ui Test 패키지 수정

chanho0908 committed:
refactor : 외부에서 호출되지 않는 Presenter Method 가시성 변경

chanho0908 commented on app/src/main/java/woowacourse/movie/view/movies/MovieListContract.kt:
```diff
@@ -7,7 +7,5 @@ interface MovieListContract {
-    interface Presenter {
-        fun loadUiData()
-    }
+    interface Presenter
```
`Presente`r에서 호출할 수 있는 초기화 로직들을
`init block`에서 호출하다보니 외부에 노출될 필요가 없다고 생각하여
오버라이드를 해제하고 `private`으로 변경했는데 이로인해 `Presenter`가 그저
마커 인터페이스가 되어버린 느낌입니다 🥲

이런 경우엔 프레젠터에 메서드를 유지하는 것이 좋을까요 ?

laco-dev reviewed (approved):
고생 많으셨습니다.

다음 미션에서도 MVP를 이어서 학습하니 연습을 충분히 해보시면 좋을 것 같아요.
너무 매몰되지 마시고 컨디션 조절 잘 해주세요.
